ouroboros-consensus/test-consensus/Test/Dynamic/Util.hs

```diff
@@ -221,34 +233,219 @@ leaderScheduleFromTrace (NumSlots numSlots) = LeaderSchedule .
+      -- For example, for a test with 4 nodes and 10 slots, the durations could
+      -- be @[0,6,1,2]@ if the first two nodes join simultaneously at slot 0,
+      -- the third joins at slot 6, the fourth joins at slot 7, the full net
+      -- runs for 2 more slots, and so the test ends at slot 9 (i.e. the 10th
+      -- slot).
+    deriving (Eq)
+
+instance Show ConnectionPlan where
+  showsPrec p connPlan@(ConnectionPlan durs) = showParen (p > 10) $
+      showString "ConnectionPlan" .
+      showSpace .
+      showsPrec 0 durs .
+      showSpace .
+      showParen True
+        ( showString "ie" .
+          showSpace .
+          showsPrec 0 interp )
+    where
+      interp = (numNodes, numSlots, map unSlotNo joinSlots)
+
+      numNodes = connectionPlanNumCoreNodes connPlan
+      (joinSlots, numSlots) = interpretConnectionPlan connPlan
+
+connectionPlanNumSlots :: ConnectionPlan -> NumSlots
+connectionPlanNumSlots (ConnectionPlan durs) = NumSlots (sum durs + 1)
+
+connectionPlanNumCoreNodes :: ConnectionPlan -> NumCoreNodes
+connectionPlanNumCoreNodes (ConnectionPlan durs) = NumCoreNodes (length durs)
+
+-- | When each node joins (Core nodes Ids @0 .. n - 1@),
+-- and number of slots in the simulation.
+interpretConnectionPlan :: ConnectionPlan -> ([SlotNo], NumSlots)
+interpretConnectionPlan (ConnectionPlan durs) = (joinSlots, numSlots)
+  where
+    (durSum, joinSlots) = mapAccumL
+        (\acc dur -> (acc + dur,SlotNo (toEnum acc)))
+        0 durs
+    numSlots = NumSlots (durSum + 1)
+
+-- | The trivial \"plan\" where all nodes join immediately in slot 0.
+trivialConnectionPlan :: NumCoreNodes -> NumSlots -> ConnectionPlan
+trivialConnectionPlan (NumCoreNodes n) (NumSlots t) =
+    ConnectionPlan $ replicate (n - 1) 0 ++ [lastSlot]
+  where
+    lastSlot = t - 1
+
+-- | Inclusive minimum bound for the last duration from 'getConnectionPlan'.
+--
+-- Must be at least 1 slot so that the last node to join has an opportunity to
+-- catch up.
+--
+-- (TODO NSF someone else please review that claim.)
+minFinalDuration :: Int
+minFinalDuration = 1   {- slots -}
+
+-- | Generate a connection plan with the given properties.
+genConnectionPlan ::
+     NumCoreNodes
+     -- ^ PREREQ: @>= 1@
+  -> NumSlots
+     -- ^ PREREQ: @>= 1@
+     --
+     -- If @== 1@, the plan is 'trivialConnectionPlan'.
+  -> Gen ConnectionPlan
+genConnectionPlan (NumCoreNodes n) (NumSlots t)
+  | n < 1 || t < 1 = error $ "Cannot generate ConnectionPlan: "
+    ++ show (NumCoreNodes n, NumSlots t)
+
+  | t < 2 = pure $
+    trivialConnectionPlan (NumCoreNodes n) (NumSlots t)
+
+  | otherwise = do
+    let lastSlot = t - 1
+
+    durs <- boundedVector (n - 1) (lastSlot - minFinalDuration)
+    let finalDur = lastSlot - sum durs
+
+    -- INVARIANT: finalDur >= minFinalDuration
+    -- INVARIANT: sum durs + finalDur = lastSlot = t - 1
+    pure $ ConnectionPlan (durs ++ [finalDur])
+  where
+    -- TODO My single experiment indicates that this could be a bit of a
```

Comment:
Hmm, do I remember correctly that this was the 30% increase in test run-time you had? 30% is too much, so we have to do something about this.

Thinking out loud:
* The *weak composition* approach you mention, does that skew the distribution such that the earlier elements in the list are larger and the later elements are smaller? If that is the only thing, then we could simply `shuffle` the list in the end (but I may be missing something).
* Could we simply *scale* the list at the end (with some ratio + rounding)?